polygon/p2p/message_listener.go

```diff
@@ -0,0 +1,141 @@
+	streamCtxCancel         context.CancelFunc
+	logger                  log.Logger
+	sentryClient            direct.SentryClient
+	observersMu             sync.Mutex
+	inboundMessageObservers map[sentry.MessageId]map[MessageObserver[*sentry.InboundMessage]]struct{}
+	stopWg                  sync.WaitGroup
+}
+
+func (ml *messageListener) Start(ctx context.Context) {
+	ml.once.Do(func() {
+		ml.streamCtx, ml.streamCtxCancel = context.WithCancel(ctx)
+		go ml.listenBlockHeaders66()
+	})
+}
+
+func (ml *messageListener) Stop() {
+	ml.streamCtxCancel()
+	ml.stopWg.Wait()
+}
+
+func (ml *messageListener) RegisterBlockHeaders66Observer(observer MessageObserver[*sentry.InboundMessage]) {
+	ml.registerInboundMessageObserver(observer, sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) UnregisterBlockHeaders66Observer(observer MessageObserver[*sentry.InboundMessage]) {
+	ml.unregisterInboundMessageObserver(observer, sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) registerInboundMessageObserver(observer MessageObserver[*sentry.InboundMessage], messageId sentry.MessageId) {
+	ml.observersMu.Lock()
+	defer ml.observersMu.Unlock()
+
+	if observers, ok := ml.inboundMessageObservers[messageId]; ok {
+		observers[observer] = struct{}{}
+	} else {
+		ml.inboundMessageObservers[messageId] = map[MessageObserver[*sentry.InboundMessage]]struct{}{
+			observer: {},
+		}
+	}
+}
+
+func (ml *messageListener) unregisterInboundMessageObserver(observer MessageObserver[*sentry.InboundMessage], messageId sentry.MessageId) {
+	ml.observersMu.Lock()
+	defer ml.observersMu.Unlock()
+
+	if observers, ok := ml.inboundMessageObservers[messageId]; ok {
+		delete(observers, observer)
+	}
+}
+
+func (ml *messageListener) listenBlockHeaders66() {
+	ml.listenInboundMessage("BlockHeaders66", sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) listenInboundMessage(name string, msgId sentry.MessageId) {
+	ml.stopWg.Add(1)
+	defer ml.stopWg.Done()
+
+	sentrymulticlient.SentryReconnectAndPumpStreamLoop(
+		ml.streamCtx,
+		ml.sentryClient,
+		ml.statusDataFactory(),
+		name,
+		ml.messageStreamFactory([]sentry.MessageId{msgId}),
+		ml.inboundMessageFactory(),
+		ml.inboundMessageHandler(),
+		nil,
+		ml.logger,
+	)
+}
+
+func (ml *messageListener) statusDataFactory() sentrymulticlient.StatusDataFactory {
+	return func() *sentry.StatusData {
+		return &sentry.StatusData{}
+	}
+}
+
+func (ml *messageListener) messageStreamFactory(ids []sentry.MessageId) sentrymulticlient.SentryMessageStreamFactory {
+	return func(streamCtx context.Context, sentryClient direct.SentryClient) (sentrymulticlient.SentryMessageStream, error) {
+		return sentryClient.Messages(streamCtx, &sentry.MessagesRequest{Ids: ids}, grpc.WaitForReady(true))
+	}
+}
+
+func (ml *messageListener) inboundMessageFactory() sentrymulticlient.MessageFactory[*sentry.InboundMessage] {
+	return func() *sentry.InboundMessage {
+		return new(sentry.InboundMessage)
```

Comment:
This looks very generic. What's the use case where we need to customize this?

Reply:
what do you mean?

Reply:
if the question is why is the MessageFactory generic - you can have a factory for PeerEvents messages too in addition to a factory for InboundMessage